src/se/terhol/pisemka33/TeachersNotebook.java

```diff
@@ -0,0 +1,99 @@
+package se.terhol.pisemka33;
+
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.SortedMap;
+import java.util.TreeMap;
+
+/**
+ * @author Tereza Holm
+ */
+public class TeachersNotebook implements ITeachersNotebook {
+    private SortedMap<Student, List<Mark>> listOfStudents = new TreeMap<>();
+
+    @Override
+    public void addMark(Student student, Mark mark) {
+        if (student == null || mark == null) {
+            throw new IllegalArgumentException();
+        }
+        if (listOfStudents.containsKey(student)) {
+            listOfStudents.get(student).add(mark);
+        } else {
+            List<Mark> marks = new ArrayList<>();
+            marks.add(mark);
+            listOfStudents.put(student, marks);
+        }
+    }
+
+    @Override
+    public List<Mark> getMarks(Student student) {
+        return listOfStudents.get(student);
+    }
+
+    @Override
+    public int getFinalMark(Student student) {
+        float sum = 0;
+        float numberOfMarks = 0;
+        float finalMark = 0;
+        if (!listOfStudents.containsKey(student)) {
+            return -1;
+        }
+        for (Mark mark : listOfStudents.get(student)) {
+            sum = sum + mark.getValue();
+            numberOfMarks++;
+        }
+        if (numberOfMarks != 0) {
+            finalMark = sum / numberOfMarks;
+        }
+        return Math.round(finalMark);
+    }
+
+    @Override
+    public void save(File file) throws IOException {
+        BufferedWriter writer = new BufferedWriter(new FileWriter(file));
+        for (Student student : listOfStudents.keySet()) {
+            writer.write(student.toString() + ":" + this.allMarks(student));
```

Comment:
It is much better to use `String.format()` to output the text instead of String concatenation. 

Material to study: https://dzone.com/articles/java-string-format-examples

Reply:
Done.